src/main/java/hudson/tasks/junit/CaseResult.java

```diff
@@ -256,10 +262,25 @@ private static String getSkippedMessage(Element testCase) {
         return message;
     }
 
-    public String getDisplayName() {
+    public String getTransformedTestName() {
         return TestNameTransformer.getTransformedName(testName);
     }
 
+    public String getDisplayName() {
+        // Only prepend the enclosing flow node names if there are any and the run this is in has multiple blocks directly containing
+        // test results.
+        if (!getEnclosingFlowNodeNames().isEmpty()) {
```

Comment:
I think so.  What we'd have to do is add a mapping from the current head(s) to the enclosing, unterminated block, and then from that block or blocks to the other enclosing blocks.  With each new head, we update that mapping. 

Would look like a mapping of nodeId -> enclosingBlock, in addition to tracking overall unclosed blocks